emasamplerate.go

```diff
@@ -0,0 +1,300 @@
+	// if it is not repeatedly observed, but will never truly reach it, so we have to decide what constitutes "zero".
+	// Keys with averages below this threshold will be removed from the EMA. Default is the same as Weight, as this prevents
+	// a key with the smallest integer value (1) from being aged out immediately.
+	AgeOutValue float64
+
+	// BurstMultiple, if set, is multiplied by the sum of the running average of counts to define
+	// the burst detection threshold. If total counts observed for a given interval exceed the threshold
+	// EMA is updated immediately, rather than waiting on the AdjustmentInterval.
+	// Defaults to 2; negative value disables. With a default of 2, if your traffic suddenly doubles,
+	// burst detection will kick in.
+	BurstMultiple float64
+
+	// BurstDetectionDelay indicates the number of intervals to run after Start is called before burst detection kicks in.
+	// Defaults to 3
+	BurstDetectionDelay uint
+
+	savedSampleRates map[string]int
+	currentCounts    map[string]float64
+	movingAverage    map[string]float64
+	burstThreshold   float64
+	currentBurstSum  float64
+	intervalCount    uint
+	burstSignal      chan struct{}
+
+	// haveData indicates that we have gotten a sample of traffic. Before we've
+	// gotten any samples of traffic, we should we should use the default goal
+	// sample rate for all events instead of sampling everything at 1
+	haveData bool
+
+	lock sync.Mutex
+
+	// used only in tests
+	testSignalMapsDone chan struct{}
+}
+
+func (e *EMASampleRate) Start() error {
+	// apply defaults
+	if e.AdjustmentInterval == 0 {
+		e.AdjustmentInterval = 15
+	}
+	if e.GoalSampleRate == 0 {
+		e.GoalSampleRate = 10
+	}
+	if e.Weight == 0 {
+		e.Weight = 0.5
+	}
+	if e.AgeOutValue == 0 {
+		e.AgeOutValue = e.Weight
+	}
+	if e.BurstMultiple == 0 {
+		e.BurstMultiple = 2
+	}
+	if e.BurstDetectionDelay == 0 {
+		e.BurstDetectionDelay = 3
+	}
+
+	e.savedSampleRates = make(map[string]int)
+	e.currentCounts = make(map[string]float64)
+	e.movingAverage = make(map[string]float64)
+	e.burstSignal = make(chan struct{})
+
+	go func() {
+		ticker := time.NewTicker(time.Second * time.Duration(e.AdjustmentInterval))
+		for {
+			select {
+			case <-e.burstSignal:
+				// reset ticker when we get a burst
+				ticker.Stop()
+				ticker = time.NewTicker(time.Second * time.Duration(e.AdjustmentInterval))
+				e.updateMaps()
+			case <-ticker.C:
+				e.updateMaps()
+				e.intervalCount++
+			}
+		}
+	}()
+	return nil
+}
+
+// updateMaps calculates a new saved rate map based on the contents of the
+// counter map
+func (e *EMASampleRate) updateMaps() {
+	// make a local copy of the sample counters for calculation
+	e.lock.Lock()
+	tmpCounts := e.currentCounts
+	e.currentCounts = make(map[string]float64)
+	e.currentBurstSum = 0
+	e.lock.Unlock()
+	// short circuit if no traffic
+	numKeys := len(tmpCounts)
+	if numKeys == 0 {
+		// no traffic the last interval, don't update anything
+		return
```

Comment:
why does the EMA not decay if there's zero traffic during the current interval?

Reply:
If there is some kind of error ingesting data for some period of time, forgetting everything in the EMA would be problematic. I can see arguments on either side of this, so maybe I could make this an option?

Reply:
hmm. I'm thinking more likely than an error triggering this would be a periodic dump of data to the dataset - ELB style where you get chunks every so often and then nothing for a while.  In that case, not continuing to decay the EMA during the silent period makes sense. 

I think you should leave it as is. Maybe add a comment saying it's intentional (since the similar code in the EMA calculation does decay silent keys when there are other keys present and the two bits will be compared even though they're serving different goals).